File: src/main/java/com/github/sarhatabaot/messages/generate/WriteJsonClass.java
package com.github.sarhatabaot.messages.generate;

import com.github.sarhatabaot.messages.model.FileType;
import com.github.sarhatabaot.messages.model.TypeKeyValue;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.google.gson.stream.JsonReader;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * @author sarhatabaot
 */
public class WriteJsonClass extends WriteClass<JsonElement>{

    public WriteJsonClass(final String targetPackage, final String basePath, final String privateConstructor, final boolean overwriteClasses) {
        super(targetPackage, basePath, privateConstructor, overwriteClasses);
    }


    @Override
    public FileType getFileType() {
        return FileType.JSON;
    }

    @Override
    public boolean isPrimitive(final @NotNull JsonElement element) {
        return element.isJsonPrimitive();
    }

    @Override
    public Set<Map.Entry<String, JsonElement>> getRootEntrySet(File file) throws FileNotFoundException {
        JsonReader reader = new JsonReader(new FileReader(file));
        JsonElement rootElement = JsonParser.parseReader(reader);
        return rootElement.getAsJsonObject().entrySet();
    }

    @Override
    public Set<Map.Entry<String, JsonElement>> getEntrySetFromValue(@NotNull JsonElement jsonElement) {
        return jsonElement.getAsJsonObject().entrySet();
    }

    @Override
    public TypeKeyValue getEntryValue(final @NotNull JsonElement value) {
        JsonPrimitive primitive = value.getAsJsonPrimitive();

        if(primitive.isNumber()) {
            try {
                return new TypeKeyValue(Double.class, String.valueOf(primitive.getAsDouble()));
            } catch (NumberFormatException e){
                return new TypeKeyValue(Integer.class, String.valueOf(primitive.getAsInt()));
            }
        }

        return new TypeKeyValue(String.class, value.getAsString());
    }

    @Override
    public boolean isArray(final @NotNull JsonElement element) {
        return element.isJsonArray();
    }

    @Override
    public String[] getAsStringArray(final @NotNull JsonElement value) {
        List<String> stringList = new ArrayList<>();
        JsonArray jsonElements = value.getAsJsonArray();
        for(JsonElement element: jsonElements) {
            stringList.add(element.getAsString());
        }
        return stringList.toArray(new String[0]);
    }
}

File: src/main/kotlin/com/github/sarhatabaot/messages/generate/WriteJsonClass.kt
package com.github.sarhatabaot.messages.generate

import com.github.sarhatabaot.messages.model.FileType
import com.github.sarhatabaot.messages.model.TypeKeyValue
import com.google.gson.JsonElement
import com.google.gson.JsonParser
import com.google.gson.stream.JsonReader
import java.io.File
import java.io.FileNotFoundException
import java.io.FileReader

/**
 *
 * @author sarhatabaot
 */
class WriteJsonClass(
    targetPackage: String,
    basePath: String,
    privateConstructor: String?,
    overwriteClasses: Boolean
): WriteClass<JsonElement>(
    targetPackage,
    basePath,
    privateConstructor,
    overwriteClasses
) {
    override fun getFileType() = FileType.JSON

    override fun isPrimitive(element: JsonElement) = element.isJsonPrimitive

    override fun isArray(element: JsonElement) = element.isJsonArray

    @Throws(FileNotFoundException::class)
    override fun getRootEntrySet(file: File?): Set<Map.Entry<String, JsonElement>> {
        val reader = JsonReader(file?.let { FileReader(it) })
        val rootElement = JsonParser.parseReader(reader)
        return rootElement.asJsonObject.entrySet()
    }

    override fun getEntrySetFromValue(value: JsonElement): Set<Map.Entry<String, JsonElement>> {
        return value.asJsonObject.entrySet()
    }

    override fun getEntryValue(value: JsonElement): TypeKeyValue {
        val primitive = value.asJsonPrimitive

        return if (primitive.isNumber) {
            try {
                TypeKeyValue(Double::class.java, primitive.asDouble.toString())
            } catch (e: NumberFormatException) {
                TypeKeyValue(Int::class.java, primitive.asInt.toString())
            }
        } else TypeKeyValue(String::class.java, value.asString)

    }

    override fun getAsStringArray(value: JsonElement): Array<String> {
        val stringList: MutableList<String> = ArrayList()
        val jsonElements = value.asJsonArray
        for (element in jsonElements) {
            stringList.add(element.asString)
        }
        return stringList.toTypedArray<String>()
    }
}